File: ios/PretextNativeModels.swift
import CoreText
import Foundation

internal struct NativeTokenDescriptor {
    let text: String
    let startUTF16: Int
    let endUTF16: Int
    let style: NativeTextStyle
    let inlineBox: NativeInlineBox?

    init(
        text: String,
        startUTF16: Int,
        endUTF16: Int,
        style: NativeTextStyle,
        inlineBox: NativeInlineBox? = nil
    ) {
        self.text = text
        self.startUTF16 = startUTF16
        self.endUTF16 = endUTF16
        self.style = style
        self.inlineBox = inlineBox
    }
}

internal struct NativePreparedToken {
    let text: String
    let startUTF16: Int
    let endUTF16: Int
    let width: Double
    let lineHeight: Double
    let ascent: Double
    let descent: Double
}

internal struct NativePreparedParagraphSeed {
    let text: String
    let tokens: [NativeTokenDescriptor]
    let breakUnits: [NativeTokenDescriptor]
    let runs: [NativeTextRun]
    let atomicSpans: [NativeAtomicSpan]
    let inlineBoxes: [NativeInlineBox]
    let textUnits: Int
    let forceTokenLayout: Bool
}

internal final class NativePreparedParagraph {
    let text: NSString
    let attributedText: NSAttributedString
    let typesetter: CTTypesetter?
    let tokens: [NativePreparedToken]
    let breakUnits: [NativePreparedToken]
    let runs: [NativeTextRun]
    let atomicSpans: [NativeAtomicSpan]
    let inlineBoxes: [NativeInlineBox]
    let forceTokenLayout: Bool

    init(
        text: String,
        attributedText: NSAttributedString,
        typesetter: CTTypesetter?,
        tokens: [NativePreparedToken],
        breakUnits: [NativePreparedToken],
        runs: [NativeTextRun],
        atomicSpans: [NativeAtomicSpan],
        inlineBoxes: [NativeInlineBox],
        forceTokenLayout: Bool
    ) {
        self.text = text as NSString
        self.attributedText = attributedText
        self.typesetter = typesetter
        self.tokens = tokens
        self.breakUnits = breakUnits
        self.runs = runs
        self.atomicSpans = atomicSpans
        self.inlineBoxes = inlineBoxes
        self.forceTokenLayout = forceTokenLayout
    }
}

internal struct NativeAtomicSpan {
    let startUTF16: Int
    let endUTF16: Int
    let source: String
}

internal struct NativeInlineBox {
    let boxId: String
    let startUTF16: Int
    let endUTF16: Int
    let width: Double
    let height: Double
    let baseline: Double
    let breakBehavior: String
    let accessibilityLabel: String?
    let accessibilityHint: String?
    let accessibilityRole: String?
}

internal final class NativePreparedCorpus {
    let paragraphs: [NativePreparedParagraph]
    let baseStyle: NativeTextStyle
    let lineHeight: Double

    private let layoutCacheLimit = 12
    private let layoutCacheLock = NSLock()
    private var layoutCache: [NativeLayoutRequest: [[NativeLineLayout]]] = [:]
    private var layoutCacheOrder: [NativeLayoutRequest] = []

    init(
        paragraphs: [NativePreparedParagraph],
        baseStyle: NativeTextStyle,
        lineHeight: Double
    ) {
        self.paragraphs = paragraphs
        self.baseStyle = baseStyle
        self.lineHeight = lineHeight
    }

    func resolveLineLayouts(
        request: NativeLayoutRequest,
        builder: () -> [[NativeLineLayout]]
    ) -> [[NativeLineLayout]] {
        layoutCacheLock.lock()
        if let cached = layoutCache[request] {
            layoutCacheOrder.removeAll(where: { $0 == request })
            layoutCacheOrder.append(request)
            layoutCacheLock.unlock()
            return cached
        }
        layoutCacheLock.unlock()

        let computed = builder()

        layoutCacheLock.lock()
        if let cached = layoutCache[request] {
            layoutCacheLock.unlock()
            return cached
        }

        layoutCache[request] = computed
        layoutCacheOrder.removeAll(where: { $0 == request })
        layoutCacheOrder.append(request)

        while layoutCacheOrder.count > layoutCacheLimit {
            let evictedRequest = layoutCacheOrder.removeFirst()
            layoutCache.removeValue(forKey: evictedRequest)
        }
        layoutCacheLock.unlock()

        return computed
    }
}

internal struct NativeLineLayout {
    let textStartUTF16: Int
    let textEndUTF16: Int
    let width: Double
    let left: Double
    let top: Double
    let height: Double
    let ascent: Double
    let descent: Double
    let layoutEngine: String
    let fallbackReason: String?
    let ctLine: CTLine?

    init(
        textStartUTF16: Int,
        textEndUTF16: Int,
        width: Double,
        left: Double,
        top: Double,
        height: Double,
        ascent: Double,
        descent: Double,
        layoutEngine: String = layoutEngineIosManualTokenFallback,
        fallbackReason: String? = fallbackReasonManualHeightEstimate,
        ctLine: CTLine? = nil
    ) {
        self.textStartUTF16 = textStartUTF16
        self.textEndUTF16 = textEndUTF16
        self.width = width
        self.left = left
        self.top = top
        self.height = height
        self.ascent = ascent
        self.descent = descent
        self.layoutEngine = layoutEngine
        self.fallbackReason = fallbackReason
        self.ctLine = ctLine
    }
}

internal enum TokenMode {
    case whitespace
    case text
}

internal struct NativeLayoutRequest: Hashable {
    let width: Double
    let left: Double
    let whiteSpace: String
    let wordBreak: String
    let shapeSlices: [NativeShapeSlice]
}

internal struct NativeShapeSlice: Hashable {
    let top: Double
    let height: Double
    let left: Double
    let width: Double
}

internal struct NativeLineConstraint {
    let left: Double
    let width: Double
}

File: android/src/main/java/com/margelo/nitro/pretext/PretextNativeModels.kt
package com.margelo.nitro.pretext

import android.text.TextPaint

internal data class NativePreparedParagraphSeed(
  val text: String,
  val tokens: List<NativeTokenDescriptor>,
  val breakUnits: List<NativeTokenDescriptor>,
  val runs: List<NativeTextRun>,
  val atomicSpans: List<NativeAtomicSpan>,
  val inlineBoxes: List<NativeInlineBox>,
  val textUnits: Int,
  val forceTokenLayout: Boolean,
)

internal data class NativeTokenDescriptor(
  val text: String,
  val start: Int,
  val end: Int,
  val style: NativeTextStyle,
  val inlineBox: NativeInlineBox? = null,
)

internal data class NativePreparedToken(
  val text: String,
  val start: Int,
  val end: Int,
  val width: Double,
  val lineHeight: Double,
  val ascent: Double,
  val descent: Double,
)

internal data class NativePreparedParagraph(
  val text: String,
  val styledText: CharSequence,
  val measuredText: Any?,
  val tokens: List<NativePreparedToken>,
  val breakUnits: List<NativePreparedToken>,
  val runs: List<NativeTextRun>,
  val atomicSpans: List<NativeAtomicSpan>,
  val inlineBoxes: List<NativeInlineBox>,
  val forceTokenLayout: Boolean,
)

internal data class NativeAtomicSpan(
  val start: Int,
  val end: Int,
  val source: String,
)

internal data class NativeInlineBox(
  val boxId: String,
  val start: Int,
  val end: Int,
  val width: Double,
  val height: Double,
  val baseline: Double,
  val breakBehavior: String,
  val accessibilityLabel: String?,
  val accessibilityHint: String?,
  val accessibilityRole: String?,
)

internal class NativePreparedCorpus(
  val paragraphs: List<NativePreparedParagraph>,
  val baseStyle: NativeTextStyle,
  val lineHeight: Double,
  val textPaint: TextPaint,
  val includeFontPadding: Boolean,
  val textDirection: ParagraphTextDirection,
  val lineBreaker: Any?,
) {
  private val layoutCacheLock = Any()
  private val layoutCache =
    LinkedHashMap<NativeLayoutRequest, List<List<NativeLineLayout>>>(16, 0.75f, true)

  fun resolveLineLayouts(
    request: NativeLayoutRequest,
    builder: () -> List<List<NativeLineLayout>>,
  ): List<List<NativeLineLayout>> {
    synchronized(layoutCacheLock) {
      layoutCache[request]?.let { return it }
    }

    val computed = builder()

    synchronized(layoutCacheLock) {
      layoutCache[request]?.let { return it }
      layoutCache[request] = computed
      while (layoutCache.size > LAYOUT_CACHE_LIMIT) {
        val iterator = layoutCache.entries.iterator()
        if (!iterator.hasNext()) {
          break
        }
        iterator.next()
        iterator.remove()
      }
    }

    return computed
  }

  private companion object {
    private const val LAYOUT_CACHE_LIMIT = 12
  }
}

internal data class NativeLineLayout(
  val textStart: Int,
  val textEnd: Int,
  val width: Double,
  val left: Double,
  val top: Double,
  val height: Double,
  val ascent: Double,
  val descent: Double,
  val layoutEngine: String = LAYOUT_ENGINE_ANDROID_LEGACY_FALLBACK,
  val fallbackReason: String? = FALLBACK_REASON_MANUAL_HEIGHT_ESTIMATE,
)

internal data class NativeLineFontPadding(
  val top: Double,
  val bottom: Double,
)

internal data class NativeLayoutRequest(
  val width: Double,
  val left: Double,
  val whiteSpace: String,
  val wordBreak: String,
  val shapeSlices: List<NativeShapeSlice>,
)

internal data class NativeShapeSlice(
  val top: Double,
  val height: Double,
  val left: Double,
  val width: Double,
)

internal data class NativeLineConstraint(
  val left: Double,
  val width: Double,
  val height: Double,
)

internal enum class TokenMode {
  WHITESPACE,
  TEXT,
}
